wp-includes/mysql/class-wp-mysql-token.php

```diff
@@ -24,6 +24,114 @@ public function get_name(): string {
 		return $name;
 	}
 
+	/**
+	 * Get the real unquoted value of the token.
+	 *
+	 * @return string The token value.
+	 */
+	public function get_value(): string {
+		$value = $this->get_bytes();
+		if (
+			WP_MySQL_Lexer::SINGLE_QUOTED_TEXT === $this->id
+			|| WP_MySQL_Lexer::DOUBLE_QUOTED_TEXT === $this->id
+			|| WP_MySQL_Lexer::BACK_TICK_QUOTED_ID === $this->id
+		) {
+			// Remove bounding quotes.
+			$quote = $value[0];
+			$value = substr( $value, 1, -1 );
+
+			/**
+			 * Unescape MySQL escape sequences.
+			 *
+			 * MySQL string literals use backslash as an escape character, and
+			 * the string bounding quotes can also be escaped by being doubled.
+			 *
+			 * The escaping is done according to the following rules:
+			 *
+			 *   1. Some special character escape sequences are recognized.
+			 *      For example, "\n" is a newline character, "\0" is ASCII NULL.
+			 *   2. A specific treatment is applied to "\%" and "\_" sequences.
+			 *      This is due to their special meaning for pattern matching.
+			 *   3. Other backslash-prefixed characters resolve to their literal
+			 *      values. For example, "\x" represents "x", "\\" represents "\".
+			 *
+			 * Despite looking similar, these rules are different from the C-style
+			 * string escaping, so we cannot use "strip(c)slashes()" in this case.
+			 *
+			 * @TODO: Handle NO_BACKSLASH_ESCAPES SQL mode.
+			 *
+			 * See: https://dev.mysql.com/doc/refman/8.4/en/string-literals.html
+			 */
+			$replacements = array(
+				/*
+				 * MySQL special character escape sequences.
+				 */
+				'\0'   => chr( 0 ),  // An ASCII NULL (X'00') character.
+				"\'"   => "'",       // A single quote (') character.
+				'\"'   => '"',       // A double quote (") character.
+				'\b'   => chr( 8 ),  // A backspace character.
+				'\n'   => "\n",      // A newline (linefeed) character.
+				'\r'   => "\r",      // A carriage return character.
+				'\t'   => "\t",      // A tab character.
+				'\Z'   => chr( 26 ), // An ASCII 26 (Control+Z) character.
+
+				/*
+				 * Normalize escaping of "%" and "_" characters.
+				 *
+				 * MySQL has unusual handling for "\%" and "\_" in all string literals.
+				 * While other sequences follow the C-style escaping ("\?" is "?", etc.),
+				 * "\%" resolves to "\%" and "\_" resolves to "\_" (unlike in C strings).
+				 *
+				 * This means that "\%" behaves like "\\%", and "\_" behaves like "\\_".
+				 * To preserve this behavior, we need to add a second backslash here.
+				 *
+				 * From https://dev.mysql.com/doc/refman/8.4/en/string-literals.html:
+				 *   > The \% and \_ sequences are used to search for literal instances
+				 *   > of % and _ in pattern-matching contexts where they would otherwise
+				 *   > be interpreted as wildcard characters. If you use \% or \_ outside
+				 *   > of pattern-matching contexts, they evaluate to the strings \% and
+				 *   > \_, not to % and _.
+				 */
+				'\%'   => '\\\\%',
```

Comment:
What do you think about using `'\\%'` instead of `\%` just to make sure PHP won't surprise us by treating the backslash as an escape sequence? Alternatively, `$backslash . '%'`?

Reply:
I thought maybe we're good since PHP docs says this:

> To specify a literal single quote, escape it with a backslash (\). To specify a literal backslash, double it (\\). All other instances of backslash will be treated as a literal backslash

But I have trust issues:

```
php > var_dump("\x5C");
string(1) "\"
```

Reply:
Ah no, that was just me using the wrong quotes:

```
php > var_dump('\x5C');
string(4) "\x5C"
php > var_dump('\n');
string(2) "\n"
```

Reply:
We're good here, but it's a good point to be a bit more explicit. I don't like the maps like `'\n' => "\n"`, where at a quick sight, the left and right sides look the same.

I followed the `$baskslash` idea in https://github.com/Automattic/sqlite-database-integration/pull/47/commits/931c82bdec4251c23338785adbf463934291eb7a, and I think it looks better now.